File: backend/src/main/java/com/nammametro/simulation/trainsim/domain/model/Disruption.java
package com.nammametro.simulation.trainsim.domain.model;

/**
 * An operational disruption injected into the simulation — a broken-down train, a blocked track,
 * station congestion, etc. {@code resourceType}/{@code resourceId} identify what it targets:
 * {@code resourceType} is fixed by {@code type} (see {@link #resourceTypeFor}), never chosen
 * independently by a caller. {@code startSeconds}/{@code durationSeconds} are simulation-elapsed
 * seconds ({@code SimulationClock#elapsedSimulationSeconds}), the same basis every other duration
 * in this engine uses.
 *
 * <p>{@code magnitudeSeconds} only matters for {@code STATION_CONGESTION} (extra dwell added to
 * every train stopping at the station while active), {@code EXTENDED_DWELL} (extra dwell added to
 * the one targeted train), and {@code CUSTOM_DELAY} (unused — that type simply freezes its target
 * train for the whole active window, same as {@code TRAIN_FAILURE}). {@code TRACK_BLOCKAGE} and
 * {@code SIGNAL_FAILURE} fully block their track for the whole active window regardless of
 * magnitude.
 */
public record Disruption(
        long id,
        DisruptionType type,
        AffectedResourceType resourceType,
        long resourceId,
        long startSeconds,
        int durationSeconds,
        int magnitudeSeconds,
        DisruptionSeverity severity,
        String description,
        DisruptionStatus status
) {

    public long endSeconds() {
        return startSeconds + durationSeconds;
    }

    public boolean isActive() {
        return status == DisruptionStatus.ACTIVE;
    }

    public Disruption withStatus(DisruptionStatus newStatus) {
        return new Disruption(id, type, resourceType, resourceId, startSeconds, durationSeconds,
                magnitudeSeconds, severity, description, newStatus);
    }

    /** The resource kind every disruption of {@code type} targets — fixed, not a free choice. */
    public static AffectedResourceType resourceTypeFor(DisruptionType type) {
        return switch (type) {
            case TRAIN_FAILURE, EXTENDED_DWELL, CUSTOM_DELAY -> AffectedResourceType.TRAIN;
            case TRACK_BLOCKAGE, SIGNAL_FAILURE -> AffectedResourceType.TRACK;
            case STATION_CONGESTION -> AffectedResourceType.STATION;
        };
    }
}

File: frontend/src/domain/trainsim/disruption.ts
export type DisruptionType =
  | "TRAIN_FAILURE"
  | "SIGNAL_FAILURE"
  | "STATION_CONGESTION"
  | "TRACK_BLOCKAGE"
  | "EXTENDED_DWELL"
  | "CUSTOM_DELAY";

export type AffectedResourceType = "TRAIN" | "TRACK" | "STATION";

export type DisruptionSeverity = "MINOR" | "MODERATE" | "MAJOR" | "SEVERE";

export type DisruptionStatus = "SCHEDULED" | "ACTIVE" | "RESOLVED" | "CANCELLED";

/** Mirrors the backend's `DisruptionResponse`. */
export interface Disruption {
  readonly id: number;
  readonly type: DisruptionType;
  readonly resourceType: AffectedResourceType;
  readonly resourceId: number;
  readonly startSeconds: number;
  readonly durationSeconds: number;
  readonly endSeconds: number;
  readonly magnitudeSeconds: number;
  readonly severity: DisruptionSeverity;
  readonly description: string;
  readonly status: DisruptionStatus;
}

/** Body for `POST /api/simulation/disruptions` — mirrors `CreateDisruptionRequest`. For
 * TRACK_BLOCKAGE/SIGNAL_FAILURE supply fromStationId/toStationId; otherwise supply resourceId. */
export interface CreateDisruptionRequest {
  readonly type: DisruptionType;
  readonly resourceId?: number;
  readonly fromStationId?: number;
  readonly toStationId?: number;
  readonly durationSeconds: number;
  readonly severity: DisruptionSeverity;
  readonly magnitudeSeconds?: number;
  readonly description?: string;
}

/** Mirrors the backend's `DisruptionAnalyticsResponse` — `GET /api/simulation/analytics/delays`. */
export interface DisruptionAnalytics {
  readonly totalDelaySeconds: number;
  readonly averageDelaySeconds: number;
  readonly maxDelaySeconds: number;
  readonly affectedTrainsCount: number;
  readonly affectedPassengers: number;
}

/** The resource kind every {@link DisruptionType} targets — fixed, mirrors `Disruption.resourceTypeFor`. */
export function resourceTypeFor(type: DisruptionType): AffectedResourceType {
  switch (type) {
    case "TRAIN_FAILURE":
    case "EXTENDED_DWELL":
    case "CUSTOM_DELAY":
      return "TRAIN";
    case "TRACK_BLOCKAGE":
    case "SIGNAL_FAILURE":
      return "TRACK";
    case "STATION_CONGESTION":
      return "STATION";
  }
}
